sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala

```diff
@@ -40,7 +41,19 @@ private[csv] object CSVInferSchema {
       csv: Dataset[String],
       caseSensitive: Boolean,
       options: CSVOptions): StructType = {
-    val firstLine: String = CSVUtils.filterCommentAndEmpty(csv, options).first()
+    val lines = CSVUtils.filterCommentAndEmpty(csv, options)
```

Comment:
You are absolutely right. Relying on exception handling is smelly, while `Option` gives more opportunities. I also see no difference from performance point of view, since both `first()` and `take(1)` call the the same function `head(1)`.